JonathanCoco/angular/product_w_factories/index.html

```diff
@@ -0,0 +1,103 @@
+        var products = [
+            {product_name: 'basketball hoop', product_price: 34.11},
+            {product_name: 'football jersey', product_price: 99.45},
+            {product_name: 'soccer goal', product_price: 250.00}];
+
+        var factory = {};
+        // Add a getStudents key to the factory object with a value of a function.
+        factory.getProducts = function (callback){
+
+            callback(products);
+        }
+
+        factory.addProduct = function(productName, productPrice, callback){
+
+            product = { product_name: productName, product_price: productPrice};
+
+            products.push(product);
+
+            callback(products);
+
+        }
+
+        factory.deleteProduct = function(index, callback){
+
+            products.splice(index, 1);
+            callback(products);
+
+        }
+        // Most important step: return the object so it can be used by the rest of our angular code
+        return factory;
+    });
+
+
+    myApp.controller('productController', ['$scope', 'productFactory', function ($scope, productFactory){
+      // we don't need a factory for this demo, it wouldn't make a difference if we did.
+        $scope.sortType     = 'product_name'; // set the default sort type
+        $scope.sortReverse  = false;  // set the default sort order
+
+        $scope.products = [];
+
+        productFactory.getProducts(function(data) {
+            $scope.products = data
+        });
+
+        $scope.addProduct = function(productName, productPrice){
+
+            productFactory.addProduct(productName, productPrice, function(data) { $scope.products = data })
+
+        }
+
+        $scope.deleteProduct = function(index){
+
+            productFactory.deleteProduct(index, function(data) { $scope.products = data })
+
+        }
+
+    }])
+  </script>
+</head>
+<body ng-app='myApp'>
+    <div ng-controller='productController'>
+      <label>Product Name:</label>
+      <input type="text" ng-model = 'productName'><br>
+
+      <label>ProductPrice:</label>
+      <input type="number" ng-model = 'productPrice'><br>
+
+      <button ng-click = "addProduct(productName, productPrice);">Add Product</button>
+
+
+      <br>
+      <br>
+      <br>
+
+      <input type="text" ng-model = 'productFilter'>
+      <br>
+
+      <table>
+          <thead>
+              <tr>
+                  <td><a href="#" ng-click="sortType = 'product_name'; sortReverse = !sortReverse">Name</a></td>
+                  <td><a href="#" ng-click="sortType = 'product_price'; sortReverse = !sortReverse">Price</a></td>
+                 <td>Actions</td>
+              <tr>
+          </thead>
+          <tr ng-repeat = "product in products | orderBy:sortType:sortReverse | filter:productFilter ">
+              <td>{{product.product_name}}</td>
+              <td>{{product.product_price | currency}}</td>
+              <td><button ng-click = "deleteProduct($index);">Delete Product</button></td>
```

Comment:
if you sort the table, does the $index stay the same?  You might want to pass in the product object to delete and when you splice the array just get the indexOf that object.